backend/stakepoold/stakepool/stakepool.go

```diff
@@ -472,6 +477,53 @@ func (spd *Stakepoold) UpdateUserDataFromMySQL() error {
 	return nil
 }
 
+// Enable early detection of outages by periodically scraping dcrd and
+// dcrwallet, and exporting the results to a monitoring system.
+func (spd *Stakepoold) ExportPrometheusMetrics(ctx context.Context, wg *sync.WaitGroup, wait time.Duration, addr string) {
+	promConnectionCount := promauto.NewGauge(prometheus.GaugeOpts{
+		Name: "dcr_peer_connections",
+		Help: "Returns a count of connections from peers",
+	})
+	promTicketsLive := promauto.NewGauge(prometheus.GaugeOpts{
+		Name: "dcr_tickets_live",
+		Help: "Returns a count of live tickets",
+	})
+	promTicketsMissed := promauto.NewGauge(prometheus.GaugeOpts{
+		Name: "dcr_tickets_missed",
+		Help: "Returns a count of missed tickets",
+	})
+	promBlockHeight := promauto.NewGauge(prometheus.GaugeOpts{
+		Name: "dcr_block_height",
+		Help: "Returns the latest block height",
+	})
+
+	go func() {
+		defer wg.Done()
+	out:
+		for {
+			select {
+			case <-ctx.Done():
+				break out
+			default:
+				dcrdConnectionCount, err := spd.NodeConnection.GetConnectionCount()
+				if err == nil {
+					promConnectionCount.Set(float64(dcrdConnectionCount))
+				}
+				stakeInfo, err := spd.GetStakeInfo()
+				if err == nil {
+					promBlockHeight.Set(float64(stakeInfo.BlockHeight))
+					promTicketsLive.Set(float64(stakeInfo.Live))
+					promTicketsMissed.Set(float64(stakeInfo.Missed))
+				}
+				time.Sleep(wait)
```

Comment:
It would be better if you made this into a case waiting for a tick on a channel from time.

Like it is now, the user will be forced to wait until this wait is over for the shutdown to happen.

If you do it like
for {
    case <-ctx.Done():
        return
    case <- time.After(wait):
        foo()
}
with no default case, then a shutdown signal will stop this loop even if in the middle of a wait.

Reply:
agreed

Reply:
Thanks.  For some reason I was expecting the for loop to run almost continuously uninhibited, eating cpu cycles and sending a cpu core to 100%.  I'm running a build with this and it seems fine though?

Reply:
Why doesn't or does?

without a default case the select should just wait and not use cpu.  What do you see?

Reply:
Oh!  I didn't realize the select was blocking if there is no default.  :mindblown: